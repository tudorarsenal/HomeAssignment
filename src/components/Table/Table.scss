.table {
  width: 100%;
  height: 100%;
  overflow-y: auto;
  border: 2px solid black;
  margin: 10px;
  padding: 5px;
  background-color: rgb(237, 237, 237);

  input {
    cursor: pointer;
  }

  .visually-hidden {
    position: absolute;
    height: 1px;
    width: 1px;
    overflow: hidden;
    clip: rect(1px, 1px, 1px, 1px);
    white-space: nowrap;
  }
}
