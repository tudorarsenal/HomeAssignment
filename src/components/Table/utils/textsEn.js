export const SELECTED = 'Selected';
export const NO_ROWS_SELECTED = 'None Selected';
export const DOWNLOAD_LABEL = 'Download Selected';
export const NO_DATA = 'No data';
export const NO_HEADERS = 'No Headers';
export const SELECT_ALL = 'Select all';
