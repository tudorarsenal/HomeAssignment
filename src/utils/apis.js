import { devices } from './mockData';

export const getDevices = () => devices;
