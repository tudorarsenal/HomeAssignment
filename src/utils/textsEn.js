export const ERROR_MESSAGE_DOWNLOAD =
  'Only available devices will be downloaded!';
export const NO_DEVICES_DOWNLOAD = 'No devices downloaded';
