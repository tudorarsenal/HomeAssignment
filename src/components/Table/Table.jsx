import { useState, useEffect } from 'react';
import PropTypes from 'prop-types';

import TableActions from './TableActions/TableActions';
import TableHeader from './TableHeader/TableHeader';
import TableBody from './TableBody/TableBody';
import { NO_HEADERS } from './utils/textsEn';

import './Table.scss';

const Table = ({
  headers,
  data,
  isSelectable,
  isDownloadable,
  preselectedRows,
  handleDownload,
  downloadRules,
}) => {
  const [selectedRows, setSelectedRows] = useState(preselectedRows);
  const [areDownloadRulesMeet, setAreDownloadRulesMeet] = useState(true);

  useEffect(() => {
    let areRulesNotMeet = false;

    // check if every selected row matches the download rule provided
    if (isDownloadable && downloadRules) {
      selectedRows?.forEach((row) => {
        const rowData = data?.find((data) => data.name === row);

        if (!downloadRules?.rule(rowData)) {
          areRulesNotMeet = true;
        }
      });
    }

    setAreDownloadRulesMeet(!areRulesNotMeet);
  }, [isDownloadable, areDownloadRulesMeet, data, downloadRules, selectedRows]);

  const areAllRowsSelected = () => selectedRows?.length === data?.length;

  const handleRowSelect = (name) => {
    if (selectedRows.indexOf(name) === -1) {
      setSelectedRows([...selectedRows, name]);
      return;
    }

    const newSelectedRows = selectedRows?.filter((row) => row !== name);
    setSelectedRows(newSelectedRows);
  };

  const handleSelectAll = () => {
    if (!areAllRowsSelected()) {
      const allRowNames = data?.map((data) => data.name);

      setSelectedRows(allRowNames);

      return;
    }

    setSelectedRows([]);
  };

  const onDownload = () => {
    if (areDownloadRulesMeet || !downloadRules.disableButton) {
      handleDownload(selectedRows);
    }
  };

  return (
    <div className="table">
      {isSelectable && (
        <TableActions
          isSelectable={isSelectable}
          isDownloadable={isDownloadable}
          selectedRowsCount={selectedRows.length}
          onSelectAll={handleSelectAll}
          onDownload={onDownload}
          areAllRowsSelected={areAllRowsSelected()}
          downloadRules={downloadRules}
          areDownloadRulesMeet={areDownloadRulesMeet}
        />
      )}
      <TableHeader headers={headers} isSelectable={isSelectable} />
      <TableBody
        data={data}
        isSelectable={isSelectable}
        onRowSelect={handleRowSelect}
        selectedRows={selectedRows}
      />
    </div>
  );
};

Table.defaultProps = {
  isSelectable: true,
  isDownloadable: false,
  preselectedRows: [],
  headers: [NO_HEADERS],
};

Table.propTypes = {
  headers: PropTypes.arrayOf(PropTypes.string),
  data: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
    })
  ).isRequired,
  isSelectable: PropTypes.bool,
  isDownloadable: PropTypes.bool,
  preselectedRows: PropTypes.arrayOf(PropTypes.string),
  handleDownload: PropTypes.func,
  downloadRules: PropTypes.shape({
    disableButton: PropTypes.bool,
    rule: PropTypes.func,
    errorMessage: PropTypes.string,
  }),
};

export default Table;
