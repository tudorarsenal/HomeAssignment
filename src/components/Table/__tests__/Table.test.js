import renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Table from '../Table';
import { devices, tableHeaders } from '../../../utils/mockData';

describe('Table', () => {
  const user = userEvent.setup();

  it('renders correctly with data', () => {
    const tree = renderer.create(<Table data={devices} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with data and headers', () => {
    const tree = renderer
      .create(<Table data={devices} headers={tableHeaders} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should call handleDownload when download button is clicked', async () => {
    const downloadRules = {
      disableButton: false,
      rule: (device) => device.status === 'available',
      errorMessage: 'Requirements not meet',
    };

    const handleDownload = jest.fn();

    render(
      <Table
        data={devices}
        headers={tableHeaders}
        isSelectable={true}
        isDownloadable={true}
        downloadRules={downloadRules}
        handleDownload={handleDownload}
      />
    );

    await user.click(screen.getByTestId('select-row-checkbox-0'));
    await user.click(screen.getByTestId('download-button'));

    expect(handleDownload).toHaveBeenCalled();
  });

  it('should select all rows when select all checkbox is clicked', async () => {
    render(<Table data={devices} headers={tableHeaders} isSelectable={true} />);

    const selectAllCheckbox = screen.getByTestId('select-all-checkbox');

    expect(selectAllCheckbox).not.toBeChecked();

    await user.click(selectAllCheckbox);

    expect(selectAllCheckbox).toBeChecked();

    for (let i = 0; i < devices.length; i++) {
      expect(screen.getByTestId(`select-row-checkbox-${i}`)).toBeChecked();
    }
  });

  it('should unselect all rows when select all checkbox is clicked a second time', async () => {
    render(<Table data={devices} headers={tableHeaders} isSelectable={true} />);

    const selectAllCheckbox = screen.getByTestId('select-all-checkbox');

    expect(selectAllCheckbox).not.toBeChecked();

    await user.click(selectAllCheckbox);

    expect(selectAllCheckbox).toBeChecked();

    await user.click(selectAllCheckbox);

    expect(selectAllCheckbox).not.toBeChecked();

    for (let i = 0; i < devices.length; i++) {
      expect(screen.getByTestId(`select-row-checkbox-${i}`)).not.toBeChecked();
    }
  });
});
