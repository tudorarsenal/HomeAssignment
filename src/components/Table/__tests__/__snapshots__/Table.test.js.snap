// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table renders correctly with data 1`] = `
<div
  className="table"
>
  <div
    className="table-actions"
  >
    <div
      className="table-actions__select"
    >
      <label
        className="visually-hidden"
        htmlFor="select-all"
      >
        Select all
      </label>
      <input
        checked={false}
        className="table-actions__check-all"
        data-testid="select-all-checkbox"
        id="select-all"
        name="select-all"
        onChange={[Function]}
        type="checkbox"
      />
      <div
        className="table-actions__selection-count"
      >
        None Selected
      </div>
    </div>
  </div>
  <div
    className="table-header table-header-selectable"
  >
    <div
      className="table-header__item table-header__no headers"
    >
      No Headers
    </div>
  </div>
  <div
    className="table-body"
  >
    <div
      className=" table-body__data-row"
    >
      <label
        className="visually-hidden"
        htmlFor="smss.exe"
      >
        smss.exe
      </label>
      <input
        checked={false}
        className="table-body__select-row"
        data-testid="select-row-checkbox-0"
        id="smss.exe"
        name="smss.exe"
        onChange={[Function]}
        type="checkbox"
      />
      <div
        className="table-body__data table-body__data-name"
      >
        smss.exe
      </div>
      <div
        className="table-body__data table-body__data-device"
      >
        Stark
      </div>
      <div
        className="table-body__data table-body__data-path"
      >
        \\Device\\HarddiskVolume2\\Windows\\System32\\smss.exe
      </div>
      <div
        className="table-body__data table-body__data-status"
      >
        scheduled
      </div>
    </div>
    <div
      className=" table-body__data-row"
    >
      <label
        className="visually-hidden"
        htmlFor="netsh.exe"
      >
        netsh.exe
      </label>
      <input
        checked={false}
        className="table-body__select-row"
        data-testid="select-row-checkbox-1"
        id="netsh.exe"
        name="netsh.exe"
        onChange={[Function]}
        type="checkbox"
      />
      <div
        className="table-body__data table-body__data-name"
      >
        netsh.exe
      </div>
      <div
        className="table-body__data table-body__data-device"
      >
        Targaryen
      </div>
      <div
        className="table-body__data table-body__data-path"
      >
        \\Device\\HarddiskVolume2\\Windows\\System32\\netsh.exe
      </div>
      <div
        className="table-body__data table-body__data-status"
      >
        <svg
          className="table-body__status-icon"
          title="status-icon-circle"
        >
          circleIcon.svg
        </svg>
        available
      </div>
    </div>
    <div
      className=" table-body__data-row"
    >
      <label
        className="visually-hidden"
        htmlFor="uxtheme.dll"
      >
        uxtheme.dll
      </label>
      <input
        checked={false}
        className="table-body__select-row"
        data-testid="select-row-checkbox-2"
        id="uxtheme.dll"
        name="uxtheme.dll"
        onChange={[Function]}
        type="checkbox"
      />
      <div
        className="table-body__data table-body__data-name"
      >
        uxtheme.dll
      </div>
      <div
        className="table-body__data table-body__data-device"
      >
        Lannister
      </div>
      <div
        className="table-body__data table-body__data-path"
      >
        \\Device\\HarddiskVolume1\\Windows\\System32\\uxtheme.dll
      </div>
      <div
        className="table-body__data table-body__data-status"
      >
        <svg
          className="table-body__status-icon"
          title="status-icon-circle"
        >
          circleIcon.svg
        </svg>
        available
      </div>
    </div>
    <div
      className=" table-body__data-row"
    >
      <label
        className="visually-hidden"
        htmlFor="cryptbase.dll"
      >
        cryptbase.dll
      </label>
      <input
        checked={false}
        className="table-body__select-row"
        data-testid="select-row-checkbox-3"
        id="cryptbase.dll"
        name="cryptbase.dll"
        onChange={[Function]}
        type="checkbox"
      />
      <div
        className="table-body__data table-body__data-name"
      >
        cryptbase.dll
      </div>
      <div
        className="table-body__data table-body__data-device"
      >
        Martell
      </div>
      <div
        className="table-body__data table-body__data-path"
      >
        \\Device\\HarddiskVolume1\\Windows\\System32\\cryptbase.dll
      </div>
      <div
        className="table-body__data table-body__data-status"
      >
        scheduled
      </div>
    </div>
    <div
      className=" table-body__data-row"
    >
      <label
        className="visually-hidden"
        htmlFor="7za.exe"
      >
        7za.exe
      </label>
      <input
        checked={false}
        className="table-body__select-row"
        data-testid="select-row-checkbox-4"
        id="7za.exe"
        name="7za.exe"
        onChange={[Function]}
        type="checkbox"
      />
      <div
        className="table-body__data table-body__data-name"
      >
        7za.exe
      </div>
      <div
        className="table-body__data table-body__data-device"
      >
        Baratheon
      </div>
      <div
        className="table-body__data table-body__data-path"
      >
        \\Device\\HarddiskVolume1\\temp\\7za.exe
      </div>
      <div
        className="table-body__data table-body__data-status"
      >
        scheduled
      </div>
    </div>
  </div>
</div>
`;

exports[`Table renders correctly with data and headers 1`] = `
<div
  className="table"
>
  <div
    className="table-actions"
  >
    <div
      className="table-actions__select"
    >
      <label
        className="visually-hidden"
        htmlFor="select-all"
      >
        Select all
      </label>
      <input
        checked={false}
        className="table-actions__check-all"
        data-testid="select-all-checkbox"
        id="select-all"
        name="select-all"
        onChange={[Function]}
        type="checkbox"
      />
      <div
        className="table-actions__selection-count"
      >
        None Selected
      </div>
    </div>
  </div>
  <div
    className="table-header table-header-selectable"
  >
    <div
      className="table-header__item table-header__name"
    >
      Name
    </div>
    <div
      className="table-header__item table-header__device"
    >
      Device
    </div>
    <div
      className="table-header__item table-header__path"
    >
      Path
    </div>
    <div
      className="table-header__item table-header__status"
    >
      Status
    </div>
  </div>
  <div
    className="table-body"
  >
    <div
      className=" table-body__data-row"
    >
      <label
        className="visually-hidden"
        htmlFor="smss.exe"
      >
        smss.exe
      </label>
      <input
        checked={false}
        className="table-body__select-row"
        data-testid="select-row-checkbox-0"
        id="smss.exe"
        name="smss.exe"
        onChange={[Function]}
        type="checkbox"
      />
      <div
        className="table-body__data table-body__data-name"
      >
        smss.exe
      </div>
      <div
        className="table-body__data table-body__data-device"
      >
        Stark
      </div>
      <div
        className="table-body__data table-body__data-path"
      >
        \\Device\\HarddiskVolume2\\Windows\\System32\\smss.exe
      </div>
      <div
        className="table-body__data table-body__data-status"
      >
        scheduled
      </div>
    </div>
    <div
      className=" table-body__data-row"
    >
      <label
        className="visually-hidden"
        htmlFor="netsh.exe"
      >
        netsh.exe
      </label>
      <input
        checked={false}
        className="table-body__select-row"
        data-testid="select-row-checkbox-1"
        id="netsh.exe"
        name="netsh.exe"
        onChange={[Function]}
        type="checkbox"
      />
      <div
        className="table-body__data table-body__data-name"
      >
        netsh.exe
      </div>
      <div
        className="table-body__data table-body__data-device"
      >
        Targaryen
      </div>
      <div
        className="table-body__data table-body__data-path"
      >
        \\Device\\HarddiskVolume2\\Windows\\System32\\netsh.exe
      </div>
      <div
        className="table-body__data table-body__data-status"
      >
        <svg
          className="table-body__status-icon"
          title="status-icon-circle"
        >
          circleIcon.svg
        </svg>
        available
      </div>
    </div>
    <div
      className=" table-body__data-row"
    >
      <label
        className="visually-hidden"
        htmlFor="uxtheme.dll"
      >
        uxtheme.dll
      </label>
      <input
        checked={false}
        className="table-body__select-row"
        data-testid="select-row-checkbox-2"
        id="uxtheme.dll"
        name="uxtheme.dll"
        onChange={[Function]}
        type="checkbox"
      />
      <div
        className="table-body__data table-body__data-name"
      >
        uxtheme.dll
      </div>
      <div
        className="table-body__data table-body__data-device"
      >
        Lannister
      </div>
      <div
        className="table-body__data table-body__data-path"
      >
        \\Device\\HarddiskVolume1\\Windows\\System32\\uxtheme.dll
      </div>
      <div
        className="table-body__data table-body__data-status"
      >
        <svg
          className="table-body__status-icon"
          title="status-icon-circle"
        >
          circleIcon.svg
        </svg>
        available
      </div>
    </div>
    <div
      className=" table-body__data-row"
    >
      <label
        className="visually-hidden"
        htmlFor="cryptbase.dll"
      >
        cryptbase.dll
      </label>
      <input
        checked={false}
        className="table-body__select-row"
        data-testid="select-row-checkbox-3"
        id="cryptbase.dll"
        name="cryptbase.dll"
        onChange={[Function]}
        type="checkbox"
      />
      <div
        className="table-body__data table-body__data-name"
      >
        cryptbase.dll
      </div>
      <div
        className="table-body__data table-body__data-device"
      >
        Martell
      </div>
      <div
        className="table-body__data table-body__data-path"
      >
        \\Device\\HarddiskVolume1\\Windows\\System32\\cryptbase.dll
      </div>
      <div
        className="table-body__data table-body__data-status"
      >
        scheduled
      </div>
    </div>
    <div
      className=" table-body__data-row"
    >
      <label
        className="visually-hidden"
        htmlFor="7za.exe"
      >
        7za.exe
      </label>
      <input
        checked={false}
        className="table-body__select-row"
        data-testid="select-row-checkbox-4"
        id="7za.exe"
        name="7za.exe"
        onChange={[Function]}
        type="checkbox"
      />
      <div
        className="table-body__data table-body__data-name"
      >
        7za.exe
      </div>
      <div
        className="table-body__data table-body__data-device"
      >
        Baratheon
      </div>
      <div
        className="table-body__data table-body__data-path"
      >
        \\Device\\HarddiskVolume1\\temp\\7za.exe
      </div>
      <div
        className="table-body__data table-body__data-status"
      >
        scheduled
      </div>
    </div>
  </div>
</div>
`;
